### PR TITLE
Java8 11 hometask

### PR DESCRIPTION
**Pull request by hontardmitry**

### CODE REVIEW

File: Java8_11/jmp-cloud-bank-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_bank_impl/CloudyBankService.java [hunk 1]
```diff
@@ -9,10 +9,8 @@
 import com.epam.jmp.dmytro_hontar.dto.bankcard.DebitBankCard;
 import com.epam.jmp.dmytro_hontar.dto.enums.BankCardType;
 
-import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
-import java.util.stream.Stream;
 
 public class CloudyBankService implements Bank {
```
Comment (Oleksandr-Zinchenko) [inline Java8_11/jmp-cloud-bank-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_bank_impl/CloudyBankService.java:15]: Please, add an empty line after the class name in future

File: Java8_11/.gitignore [hunk 2]
```diff
@@ -0,0 +1,38 @@
+target/
```
Comment (Oleksandr-Zinchenko) [inline Java8_11/.gitignore:1]: We need just one .gitignore and need to exclude idea generated files

File: Java8_11/jmp-cloud-bank-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_bank_impl/CloudyBankService.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import com.epam.jmp.dmytro_hontar.dto.enums.BankCardType;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CloudyBankService implements Bank {
+    @Override
+    public List<BankCard> createBankCards(List<User> users, BankCardType cardType) {
+        return users.stream()
+                .map((CREDIT.equals(cardType)? CreditBankCard::new : DebitBankCard::new))
```
Comment (Oleksandr-Zinchenko) [inline Java8_11/jmp-cloud-bank-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_bank_impl/CloudyBankService.java:19]: Create a map with the key - BankCarType and the value - credit cart supplier

File: Java8_11/jmp-cloud-service-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_service_impl/CloudyService.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CloudyService implements Service {
+    public CloudyService(List<User> users) {
+        this.users = users;
+    }
+
+    private final List<User> users;
+    private List<Subscription> subscriptions = new ArrayList<>();
```
Comment (Oleksandr-Zinchenko) [inline Java8_11/jmp-cloud-service-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_service_impl/CloudyService.java:20]: Add an empty line after variables

File: Java8_11/jmp-dto/src/main/java/com/epam/jmp/dmytro_hontar/dto/User.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import java.time.LocalDate;
+
+public class User {
+    public User(String name, String surname, LocalDate birthday) {
+        this.name = name;
+        this.surname = surname;
+        this.birthday = birthday;
+    }
+
+    private String name;
```
Comment (Oleksandr-Zinchenko) [inline Java8_11/jmp-dto/src/main/java/com/epam/jmp/dmytro_hontar/dto/User.java:12]: Make these fields final if you are using constructor + getter

File: Java8_11/jmp-dto/src/main/java/com/epam/jmp/dmytro_hontar/dto/bankcard/BankCard.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import com.epam.jmp.dmytro_hontar.dto.User;
+
+import java.math.BigDecimal;
+
+public abstract class BankCard {
+    private final String number;
+    private final User user;
+    private BigDecimal amount;
+
+    public BankCard(User user) {
```
Comment (Oleksandr-Zinchenko) [inline Java8_11/jmp-dto/src/main/java/com/epam/jmp/dmytro_hontar/dto/bankcard/BankCard.java:14]: Try to choose waht to use, setters + getters or constructor + getter. Don't make it messy

File: Java8_11/jmp-cloud-service-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_service_impl/CloudyService.java (hunk 4 above)
Comment (hontardmitry) [reply Java8_11/jmp-cloud-service-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_service_impl/CloudyService.java:20]: need to talk regarding the subscriptions variable

File: Java8_11/jmp-cloud-bank-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_bank_impl/CloudyBankService.java (hunk 3 above)
Comment (hontardmitry) [reply Java8_11/jmp-cloud-bank-impl/src/main/java/com/epam/jmp/dmytro_hontar/cloud_bank_impl/CloudyBankService.java:19]: What do you think if I do implement not exactly this
`    public List<BankCard> createBankCards(List<User> users, BankCardType cardType) {
        Map<BankCardType, Function<User, BankCard>> bankCardSupplier = Map.of(
            CREDIT, CreditBankCard::new,
            DEBIT, DebitBankCard::new);

        return users.stream()
                .map(bankCardSupplier.get(cardType))
                .collect(Collectors.toList());
    }`
    but this
    `public List<BankCard> createBankCards(List<User> users, BankCardType cardType) {
        return users.stream()
                .map(cardType::createCard)
                .collect(Collectors.toList());
    }`
With modified enum:
`public enum BankCardType {
    CREDIT(CreditBankCard::new),
    DEBIT(DebitBankCard::new);

    BankCardType(Function<User, BankCard> cardFunction) {
        this.cardFunction = cardFunction;
    }

    private final Function<User, BankCard> cardFunction;

    public BankCard createCard(User user) {
        return cardFunction.apply(user);
    }
}`

File: Java8_11/jmp-dto/src/main/java/com/epam/jmp/dmytro_hontar/dto/enums/BankCardType.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+public enum BankCardType {
+    CREDIT(CreditBankCard::new),
+    DEBIT(DebitBankCard::new);
+
+    BankCardType(Function<User, BankCard> cardFunction) {
+        this.cardFunction = cardFunction;
+    }
+
+    private final Function<User, BankCard> cardFunction;
```
Comment (Oleksandr-Zinchenko) [inline Java8_11/jmp-dto/src/main/java/com/epam/jmp/dmytro_hontar/dto/enums/BankCardType.java:18]: What's the point of function if you don't use the input parameter?

**Pull request merged**